proposals/time-stamp-counter-support.md

```diff
@@ -0,0 +1,274 @@
+compare from similar clock sources. For very short measurements, consider
+pinning to a single core.
+
+Due to these hazards, `time.perf_counter_ns` should likely not be backed by the TSC
+by default, but perhaps a compile-time flag could enable it for those who
+are aware of the hazards and have mitigated them.
+
+## Proposal
+
+### A place for architecture specific features
+
+The way to determine the TSC frequency on Intel CPUs is via the `cpuid`
+instruction, which is decidedly not portable. There currently isn't a good
+place for non-CPU-agnostic code to go, and as a result this proposal includes
+the creation of a `sys._arch` module, containing `sys._arch.x86_64`,
+`sys.arch.aarch64`, and future supported architectures. This balances the need to
+use architecture specific inline assembly to expose some features that
+developers may find valuable with the need to maintain a level of portability.
+
+This _arch module would include eventually implementations of runtime CPU
+feature discovery, raw time stamp counter reads, and the required
+implementations for atomics, including atomics which may not exactly fit under
+the C11 memory model. This module SHOULD NOT be used for anything which can be
+implemented on other architectures, for instance, x86's `popcnt` can be
+implemented as `vcnt.8` with ARM NEON or with a for loop, so it shouldn't be
+there. In other words, every single function in this module and it's submodules
+will have a constraint requiring a specific architecture.
+
+As we want Mojo to be as portable as possible, the module shall be hidden by
+default and the documentation of this module should discourage end-users using
+its functions unless there are no other options.
+
+### The creation of time._tsc
+
+The time module is the most logical place for most of this functionality to
+live, since they are time stamp counters. In this module, different
+architectures will be unified into a single portable API consisting of the
+following:
+
+```mojo
+@value
+struct TscFrequency:
+    """
+    Contains the TSC frequency with architecture-defined units. 
+    """
+    var _tsc_frequency: UInt64
+
+    fn get_frequency_hz(borrowed self) -> UInt64:
+        """
+        Get the frequency in Hertz. 
+        """
+        pass
+
+@value
+struct TscTimestamp:
+    """
+    Encapsulates the value of a TSC read. Units are architecture dependent.
+
+    Implementation detail: Can only reliably hold up to 2^64-1 nanoseconds or ~580 years on most architectures. 
+    """
+    var _tsc_value: UInt64
+
+    fn time_since_ns(borrowed self, borrowed other: Self, borrowed freq: TscFrequency) -> Int64:
+        """
+        Returns the time since the previous timestamp in nanoseconds using the provided frequency reference.
+        """
+        ...
+
+    fn get_raw(borrowed self) -> UInt64:
+        """
+        Access the raw, architecture dependent value of the TSC. 
+        """
+        return self._tsc_value
+
+struct Tsc:
```

Comment:
**Suggestion** Let's spell this out fully for readability sake: `TimestampCounter` (from internal review)

Reply:
ACK